main/lsp/LSPIndexer.cc

```diff
@@ -144,26 +90,79 @@ vector<core::FileHash> LSPIndexer::computeFileHashes(const vector<shared_ptr<cor
+bool LSPIndexer::canTakeFastPath(const std::vector<std::shared_ptr<core::File>> &changedFiles,
+                                 bool containsPendingTypecheckUpdates) const {
+    Timer timeit(config->logger, "fast_path_decision");
+    auto &logger = *config->logger;
+    logger.debug("Trying to see if fast path is available after {} file changes", changedFiles.size());
+    if (config->disableFastPath) {
+        logger.debug("Taking slow path because fast path is disabled.");
+        prodCategoryCounterInc("lsp.slow_path_reason", "fast_path_disabled");
+        return false;
+    }
+
+    const UnorderedMap<int, shared_ptr<core::File>> emptyMap;
+    const UnorderedMap<int, shared_ptr<core::File>> &evictedFilesRef =
+        containsPendingTypecheckUpdates ? evictedFiles : emptyMap;
+    for (auto &f : changedFiles) {
+        auto fref = initialGS->findFileByPath(f->path());
+        if (!fref.exists()) {
+            logger.debug("Taking slow path because {} is a new file", f->path());
+            prodCategoryCounterInc("lsp.slow_path_reason", "new_file");
+            return false;
+        } else {
+            const auto &oldFile = getOldFile(fref, *initialGS, evictedFilesRef);
+            ENFORCE(oldFile.getFileHash() != nullptr);
+            ENFORCE(f->getFileHash() != nullptr);
+            auto oldHash = *oldFile.getFileHash();
+            auto newHash = *f->getFileHash();
+            ENFORCE(oldHash.definitions.hierarchyHash != core::GlobalStateHash::HASH_STATE_NOT_COMPUTED);
+            if (newHash.definitions.hierarchyHash == core::GlobalStateHash::HASH_STATE_INVALID) {
+                logger.debug("Taking slow path because {} has a syntax error", f->path());
+                prodCategoryCounterInc("lsp.slow_path_reason", "syntax_error");
+                return false;
+            } else if (newHash.definitions.hierarchyHash != core::GlobalStateHash::HASH_STATE_INVALID &&
+                       newHash.definitions.hierarchyHash != oldHash.definitions.hierarchyHash) {
+                logger.debug("Taking slow path because {} has changed definitions", f->path());
+                prodCategoryCounterInc("lsp.slow_path_reason", "changed_definition");
+                return false;
+            }
+        }
+    }
+
+    logger.debug("Taking fast path");
+    return true;
+}
+
+bool LSPIndexer::canTakeFastPath(const LSPFileUpdates &edit, bool containsPendingTypecheckUpdates) const {
+    auto &logger = *config->logger;
+    // Path taken after the first time an update has been encountered. Hack since we can't roll back new files just yet.
+    if (edit.hasNewFiles) {
+        logger.debug("Taking slow path because update has a new file");
+        prodCategoryCounterInc("lsp.slow_path_reason", "new_file");
+        return false;
+    }
+    return canTakeFastPath(edit.updatedFiles, containsPendingTypecheckUpdates);
+}
+
+bool LSPIndexer::canTakeFastPath(const std::vector<std::shared_ptr<core::File>> &changedFiles) const {
```

Comment:
The only public interface for `canTakeFastPath`. This method is called in `SorbetWorkspaceEditParams`, which tells the scheduler if typechecking the edit can preempt a slow path. I defined it separately so that we don't expose the `containsPendingTypecheckUpdates` option, which is only relevant to callsites in `commitEdit`.